File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: Optional[ListNode]
        :type n: int
        :rtype: Optional[ListNode]
        """
        dummyHead =ListNode(0,head);
        first=dummyHead
        second =dummyHead

        for i in range(n+1):
            first = first.next
        while first :
            second = second.next
            first = first.next
        second.next = second.next.next
        return dummyHead.next

    
        

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *slow = head ;
        ListNode *trailSlow = new ListNode(0,head);
        ListNode *fast = head;
        for(int i = 0 ; i < n ; i++ )
            fast = fast->next;

        while(fast != NULL){
            fast = fast->next;
            slow = slow->next;
            trailSlow = trailSlow->next;
        }
        if(slow == head){
            head = head->next;
        }else{
            trailSlow->next= slow->next;

        }
        delete slow;
        slow = NULL;

        return head;
    }
};